src/main/scala/com/databricks/labs/overwatch/utils/UpgradeV2.scala

```diff
@@ -0,0 +1,405 @@
+package com.databricks.labs.overwatch.utils
+
+import com.databricks.dbutils_v1.DBUtilsHolder.dbutils
+import com.databricks.labs.overwatch.env.Workspace
+import com.databricks.labs.overwatch.pipeline.TransformFunctions._
+import com.databricks.labs.overwatch.pipeline._
+import com.databricks.labs.overwatch.utils.Helpers.{fastDrop, parOptimize}
+import com.databricks.labs.overwatch.utils.Upgrade.spark
+import org.apache.log4j.{Level, Logger}
+import org.apache.spark.sql.DataFrame
+import org.apache.spark.sql.functions.col
+import spark.implicits._
+
+import java.util.concurrent.ConcurrentHashMap
+import scala.collection.JavaConverters.mapAsScalaConcurrentMapConverter
+import scala.collection.concurrent
+import scala.collection.mutable.ArrayBuffer
+
+class UpgradeTo0610(prodWorkspace: Workspace, startStep: Int = 1, endStep: Int = 3, enableUpgradeBelowDBR104: Boolean = false) extends UpgradeHandler {
+
+
+  def this(workspace: Workspace) {
+    this(workspace, 1, 3,false)
+  }
+
+
+  override def upgrade(): DataFrame = {
+    val tempDir = s"/tmp/overwatch/upgrade0610_status__ctrl_0x111/${System.currentTimeMillis()}"
+    val etlDatabaseName = prodWorkspace.database.getDatabaseName
+    dbutils.fs.mkdirs(tempDir) // init tempDir -- if no errors it wouldn't be created
+    val blankConfig = new Config()
+    val currentSchemaVersion = SchemaTools.getSchemaVersion(etlDatabaseName)
+    val numericalSchemaVersion = getNumericalSchemaVersion(currentSchemaVersion)
+    val targetSchemaVersion = "0.610"
+    validateSchemaUpgradeEligibility(currentSchemaVersion, targetSchemaVersion)
+    validateNumericalSchemaVersion(numericalSchemaVersion, 600, 610)
+    val upgradeStatus: ArrayBuffer[UpgradeReport] = ArrayBuffer()
+    val dbrVersion = spark.conf.get("spark.databricks.clusterUsageTags.effectiveSparkVersion")
+    val dbrMajorV = dbrVersion.split("\\.").head
+    val dbrMinorV = dbrVersion.split("\\.")(1)
+    val dbrVersionNumerical = s"$dbrMajorV.$dbrMinorV".toDouble
+    val initialSourceVersions: concurrent.Map[String, Long] = new ConcurrentHashMap[String, Long]().asScala
+    val packageVersion = blankConfig.getClass.getPackage.getImplementationVersion.replaceAll("\\.", "").tail.toInt
+    val startingSchemaVersion = SchemaTools.getSchemaVersion(etlDatabaseName).split("\\.").takeRight(1).head.toInt
+    validateSchemaAndPackageVersion(startingSchemaVersion, packageVersion, 600, 610)
+
+    (startStep to endStep).foreach {
+      case 1 =>
+        val stepMsg = Some("Step 1: Upgrade Schema - Job Status Silver")
```

Comment:
I want to try and clean this up so that all upgrades ultimately submit a map of step number and upgradeStepFunction to the UpgradeHandler. Hopefully, it can look something like

```scala
private def 0610Step1(...) = ???
private def 0610Step2(...) = ???

val upgradeMap(stepNumber -> upgradeFunc, ...)
upgrade(upgradeMap)
```
something like this. The above is just pseudo code obviously but hopefully that is clear enough. The idea being that the dev creates n functions and then passes those to the upgrade -- I think this will make the code much cleaner and easier to navigate, thoughts?

Reply:
Inside the upgrade handler all the exception handlers should be obfuscated there meaning all the try / catch / finally stuff should be in the abstract class or executor not in the actual upgrade script.